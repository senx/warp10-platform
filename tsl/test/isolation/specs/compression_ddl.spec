setup
{
    CREATE TABLE ts_device_table(time INTEGER, device INTEGER, location INTEGER, value INTEGER);
    SELECT create_hypertable('ts_device_table', 'time', chunk_time_interval => 10);
    INSERT INTO ts_device_table SELECT generate_series(0,29,1), 1, 100, 20;
    ALTER TABLE ts_device_table set(timescaledb.compress, timescaledb.compress_segmentby='location', timescaledb.compress_orderby='time');
    CREATE FUNCTION lock_chunktable( name text) RETURNS void AS $$
    BEGIN EXECUTE format( 'lock table %s IN SHARE MODE', name);
    END; $$ LANGUAGE plpgsql;

}
teardown
{
   DROP TABLE ts_device_table cascade;
   DROP FUNCTION lock_chunktable( text );
}

session "I"
step "Ib"   { BEGIN; SET LOCAL lock_timeout = '50ms'; SET LOCAL deadlock_timeout = '10ms';}
step "I1"   { BEGIN; INSERT INTO ts_device_table VALUES (1, 1, 100, 100); }
step "Ic"   { COMMIT; }

session "S"
step "St" {BEGIN;} 
step "S1" { SELECT count(*) from ts_device_table; }
step "SC1" { SELECT count(*) from _timescaledb_internal._hyper_1_1_chunk; }
step "SH" { SELECT total_chunks, number_compressed_chunks from timescaledb_information.compressed_hypertable_stats where hypertable_name::text like 'ts_device_table'; }
step "Sc" {COMMIT;}

session "LC"
step "LockChunk1"   { BEGIN; select lock_chunktable (q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; }
step "UnlockChunk" {ROLLBACK;}

session "C"
step "C1"   { BEGIN; SELECT compress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; }
step "Cc"   { COMMIT; }

session "A"
step "A1" { BEGIN; ALTER TABLE ts_device_table SET ( fillfactor = 80); }
step "A2" { COMMIT; }

session "D"
step "D1"   { BEGIN; SELECT decompress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; }
step "Dc"   { COMMIT; }

#if insert in progress, compress  is blocked
permutation "LockChunk1" "I1" "C1" "UnlockChunk" "Ic" "Cc" "SC1" "S1" 

#Compress in progress, insert is blocked
permutation "LockChunk1" "C1" "I1" "UnlockChunk" "Cc" "Ic"  

#if ddl in progress, compress_chunk blocked
permutation "LockChunk1" "A1" "C1" "UnlockChunk" "Cc" "A2"
permutation "LockChunk1" "A1" "C1" "UnlockChunk" "A2" "Cc"

#concurrent compress/deocmpress on same chunk errors
permutation "LockChunk1" "C1" "D1" "UnlockChunk" "Cc" "Dc"

#concurrent compress and select should execute concurrently
permutation "LockChunk1" "C1" "S1" "UnlockChunk" "Cc" "SH" 
permutation "LockChunk1" "C1" "S1" "UnlockChunk" "SH" "Cc"
