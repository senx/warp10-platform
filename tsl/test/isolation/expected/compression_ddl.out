Parsed test spec with 6 sessions

starting permutation: LockChunk1 I1 C1 UnlockChunk Ic Cc SC1 S1
step LockChunk1: BEGIN; select lock_chunktable (q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q;
lock_chunktable

               
step I1: BEGIN; INSERT INTO ts_device_table VALUES (1, 1, 100, 100); <waiting ...>
step C1: BEGIN; SELECT compress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; <waiting ...>
step UnlockChunk: ROLLBACK;
step I1: <... completed>
step Ic: COMMIT;
step C1: <... completed>
compress_chunk 

_timescaledb_internal._hyper_1_1_chunk
step Cc: COMMIT;
step SC1: SELECT count(*) from _timescaledb_internal._hyper_1_1_chunk;
count          

0              
step S1: SELECT count(*) from ts_device_table;
count          

31             

starting permutation: LockChunk1 C1 I1 UnlockChunk Cc Ic
step LockChunk1: BEGIN; select lock_chunktable (q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q;
lock_chunktable

               
step C1: BEGIN; SELECT compress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; <waiting ...>
step I1: BEGIN; INSERT INTO ts_device_table VALUES (1, 1, 100, 100); <waiting ...>
step UnlockChunk: ROLLBACK;
step C1: <... completed>
compress_chunk 

_timescaledb_internal._hyper_3_5_chunk
step Cc: COMMIT;
step I1: <... completed>
error in steps Cc I1: ERROR:  insert/update/delete not permitted on chunk "_hyper_3_5_chunk"
step Ic: COMMIT;

starting permutation: LockChunk1 A1 C1 UnlockChunk Cc A2
step LockChunk1: BEGIN; select lock_chunktable (q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q;
lock_chunktable

               
step A1: BEGIN; ALTER TABLE ts_device_table SET ( fillfactor = 80); <waiting ...>
step C1: BEGIN; SELECT compress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; <waiting ...>
step UnlockChunk: ROLLBACK;
step A1: <... completed>
step C1: <... completed>
ERROR:  canceling statement due to user request
step Cc: COMMIT;
step A2: COMMIT;

starting permutation: LockChunk1 A1 C1 UnlockChunk A2 Cc
step LockChunk1: BEGIN; select lock_chunktable (q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q;
lock_chunktable

               
step A1: BEGIN; ALTER TABLE ts_device_table SET ( fillfactor = 80); <waiting ...>
step C1: BEGIN; SELECT compress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; <waiting ...>
step UnlockChunk: ROLLBACK;
step A1: <... completed>
step A2: COMMIT;
step C1: <... completed>
compress_chunk 

_timescaledb_internal._hyper_7_13_chunk
step Cc: COMMIT;

starting permutation: LockChunk1 C1 D1 UnlockChunk Cc Dc
step LockChunk1: BEGIN; select lock_chunktable (q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q;
lock_chunktable

               
step C1: BEGIN; SELECT compress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; <waiting ...>
step D1: BEGIN; SELECT decompress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q;
ERROR:  chunk "_hyper_9_17_chunk" is not compressed
step UnlockChunk: ROLLBACK;
step C1: <... completed>
compress_chunk 

_timescaledb_internal._hyper_9_17_chunk
step Cc: COMMIT;
step Dc: COMMIT;

starting permutation: LockChunk1 C1 S1 UnlockChunk Cc SH
step LockChunk1: BEGIN; select lock_chunktable (q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q;
lock_chunktable

               
step C1: BEGIN; SELECT compress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; <waiting ...>
step S1: SELECT count(*) from ts_device_table;
count          

30             
step UnlockChunk: ROLLBACK;
step C1: <... completed>
compress_chunk 

_timescaledb_internal._hyper_11_21_chunk
step Cc: COMMIT;
step SH: SELECT total_chunks, number_compressed_chunks from timescaledb_information.compressed_hypertable_stats where hypertable_name::text like 'ts_device_table';
total_chunks   number_compressed_chunks

3              1              

starting permutation: LockChunk1 C1 S1 UnlockChunk SH Cc
step LockChunk1: BEGIN; select lock_chunktable (q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q;
lock_chunktable

               
step C1: BEGIN; SELECT compress_chunk(q.chname) from (select ch.schema_name || '.' || ch.table_name as chname from _timescaledb_catalog.hypertable ht, _timescaledb_catalog.chunk ch where ch.hypertable_id = ht.id and ht.table_name like 'ts_device_table' order by ch.id limit 1 ) q; <waiting ...>
step S1: SELECT count(*) from ts_device_table;
count          

30             
step UnlockChunk: ROLLBACK;
step C1: <... completed>
compress_chunk 

_timescaledb_internal._hyper_13_25_chunk
step SH: SELECT total_chunks, number_compressed_chunks from timescaledb_information.compressed_hypertable_stats where hypertable_name::text like 'ts_device_table';
total_chunks   number_compressed_chunks

step Cc: COMMIT;
